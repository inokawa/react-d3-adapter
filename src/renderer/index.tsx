import React, { useLayoutEffect, useRef } from "react";
import renderer from "./renderer";

export const Renderer = (props: { children: React.ReactNode }) => {
  useLayoutEffect(() => {
    renderer.render(props.children, ref.current!);
  }, [props.children]);
  const ref = useRef<HTMLDivElement>(null);
  return <div ref={ref} />;
};
