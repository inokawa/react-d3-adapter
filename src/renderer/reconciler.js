import ReactReconciler from "react-reconciler";
import * as d3 from "d3";

const applyD3Props = (sel, el, props) => {
  Object.keys(props).forEach((k) => {
    if (k === "children") return;
    if (k.startsWith("on") && typeof props[k] === "function") {
      d3.select(el).on(k.slice(2).toLowerCase(), props[k]);
    } else {
      if (k === "className") {
        el[k] = props[k];
      } else if (k === "style") {
        Object.keys(props.style).forEach((k) => {
          sel.style(k, props.style[k]);
        });
      } else {
        sel.attr(k, props[k]);
      }
    }
  });
};

export const reconciler = ReactReconciler({
  supportsMutation: true,

  createInstance(type, props, rootContainer, hostContext, internalHandle) {
    const el = document.createElement(type);

    return { el, props, children: [] };
  },
  createTextInstance(text, rootContainer, hostContext, internalHandle) {
    return { el: document.createTextNode(text) };
  },

  appendChildToContainer(container, child) {
    const sel = d3.select(container).append(() => child.el);
    if (child.el.nodeType === 1) {
      sel.style("opacity", 0).transition().style("opacity", 1);
      applyD3Props(sel, child.el, child.props);
    }
  },
  appendChild(parent, child) {
    const sel = d3.select(parent.el).append(() => child.el);
    if (child.el.nodeType === 1) {
      sel.style("opacity", 0).transition().style("opacity", 1);
      applyD3Props(sel, child.el, child.props);
    }
  },
  appendInitialChild(parent, child) {
    const sel = d3.select(parent.el).append(() => child.el);
    if (child.el.nodeType === 1) {
      sel.style("opacity", 0).transition().style("opacity", 1);
      applyD3Props(sel, child.el, child.props);
    }
  },

  removeChildFromContainer(container, child) {
    d3.select(child.el).transition().style("opacity", 0).remove();
  },
  removeChild(parent, child) {
    d3.select(child.el).transition().style("opacity", 0).remove();
  },
  insertInContainerBefore(container, child, before) {
    const sel = d3.select(container).insert(
      () => child.el,
      () => before.el
    );
    if (child.el.nodeType === 1) {
      sel.style("opacity", 0).transition().style("opacity", 1);
      applyD3Props(sel, child.el, child.props);
    }
  },
  insertBefore(parent, child, before) {
    const sel = d3.select(parent.el).insert(
      () => child.el,
      () => before.el
    );
    if (child.el.nodeType === 1) {
      sel.style("opacity", 0).transition().style("opacity", 1);
      applyD3Props(sel, child.el, child.props);
    }
  },

  clearContainer(container) {
    let c = container.firstChild;
    while (c) {
      container.removeChild(c);
      c = container.firstChild;
    }
  },

  prepareUpdate(
    instance,
    type,
    oldProps,
    newProps,
    rootContainer,
    hostContext
  ) {
    const payload = {};
    Object.keys({ ...oldProps, ...newProps }).forEach((k) => {
      if (oldProps[k] !== newProps[k]) payload[k] = newProps[k];
    });
    return Object.keys(payload).length === 0 ? null : payload;
  },
  commitUpdate(
    instance,
    updatePayload,
    type,
    oldProps,
    newProps,
    finishedWork
  ) {
    if (instance.el.nodeType === 1) {
      const sel = d3.select(instance.el).transition();
      applyD3Props(sel, instance.el, updatePayload);
    }
  },
  commitTextUpdate(instance, oldText, newText) {
    if (oldText !== newText) instance.el.textContent = newText;
  },

  finalizeInitialChildren(instance, type, props, rootContainer, hostContext) {
    return false;
  },
  commitMount(instance, type, props, internalHandle) {},

  getRootHostContext(rootContainer) {
    return null;
  },
  getChildHostContext(parentHostContext, type, rootContainer) {
    return parentHostContext;
  },
  getPublicInstance(instance) {
    return instance;
  },
  prepareForCommit(containerInfo) {
    return null;
  },
  resetAfterCommit(containerInfo) {},
  shouldSetTextContent(type, props) {
    return false;
  },
  resetTextContent(instance) {},
});
