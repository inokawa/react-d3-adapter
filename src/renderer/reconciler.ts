// @ts-expect-error
import ReactReconciler from "react-reconciler";
import * as d3 from "d3";

type Container = HTMLElement;

type Type = string; // TODO

type Instance = {
  el: HTMLElement;
  props: any;
  children: [];
};

type TextInstance = { el: Text };

type UpdatePayload = {
  [key: string]: any;
};

const applyD3Props = (sel: any, el: HTMLElement, props: UpdatePayload) => {
  for (const k of Object.keys(props)) {
    if (k === "children") continue;
    if (k.startsWith("on") && typeof props[k] === "function") {
      d3.select(el).on(k.slice(2).toLowerCase(), props[k]);
    } else {
      if (k === "className") {
        el[k] = props[k];
      } else if (k === "style") {
        Object.keys(props.style).forEach((k) => {
          sel.style(k, props.style[k]);
        });
      } else {
        sel.attr(k, props[k]);
      }
    }
  }
};

export const reconciler = ReactReconciler({
  supportsMutation: true,

  createInstance(
    type: Type,
    props: any,
    rootContainer: Container,
    hostContext: any,
    internalHandle: any
  ): Instance {
    const el = document.createElement(type);

    return { el, props, children: [] };
  },
  createTextInstance(
    text: string,
    rootContainer: Container,
    hostContext: any,
    internalHandle: any
  ): TextInstance {
    return { el: document.createTextNode(text) };
  },

  appendChildToContainer(container: Container, child: Instance) {
    const sel = d3.select(container).append(() => child.el);
    if (child.el.nodeType === 1) {
      sel.style("opacity", 0).transition().style("opacity", 1);
      applyD3Props(sel, child.el, child.props);
    }
  },
  appendChild(parent: Instance, child: Instance) {
    const sel = d3.select(parent.el).append(() => child.el);
    if (child.el.nodeType === 1) {
      sel.style("opacity", 0).transition().style("opacity", 1);
      applyD3Props(sel, child.el, child.props);
    }
  },
  appendInitialChild(parent: Instance, child: Instance) {
    const sel = d3.select(parent.el).append(() => child.el);
    if (child.el.nodeType === 1) {
      sel.style("opacity", 0).transition().style("opacity", 1);
      applyD3Props(sel, child.el, child.props);
    }
  },

  insertInContainerBefore(
    container: Container,
    child: Instance,
    before: Instance
  ) {
    const sel = d3.select(container).insert(
      () => child.el,
      () => before.el
    );
    if (child.el.nodeType === 1) {
      sel.style("opacity", 0).transition().style("opacity", 1);
      applyD3Props(sel, child.el, child.props);
    }
  },
  insertBefore(parent: Instance, child: Instance, before: Instance) {
    const sel = d3.select(parent.el).insert(
      () => child.el,
      () => before.el
    );
    if (child.el.nodeType === 1) {
      sel.style("opacity", 0).transition().style("opacity", 1);
      applyD3Props(sel, child.el, child.props);
    }
  },

  prepareUpdate(
    instance: Instance,
    type: Type,
    oldProps: any,
    newProps: any,
    rootContainer: Container,
    hostContext: any
  ): UpdatePayload | null {
    const payload: UpdatePayload = {};
    for (const k of Object.keys({ ...oldProps, ...newProps })) {
      if (oldProps[k] !== newProps[k]) payload[k] = newProps[k];
    }
    return Object.keys(payload).length === 0 ? null : payload;
  },
  commitUpdate(
    instance: Instance,
    updatePayload: UpdatePayload,
    type: Type,
    oldProps: any,
    newProps: any,
    finishedWork: any
  ) {
    if (instance.el.nodeType === 1) {
      const sel = d3.select(instance.el).transition();
      applyD3Props(sel, instance.el, updatePayload);
    }
  },
  commitTextUpdate(instance: TextInstance, oldText: string, newText: string) {
    if (oldText !== newText) {
      // TODO textTween
      instance.el.textContent = newText;
    }
  },

  removeChildFromContainer(container: Container, child: Instance) {
    d3.select(child.el).transition().style("opacity", 0).remove();
  },
  removeChild(parent: Instance, child: Instance) {
    d3.select(child.el).transition().style("opacity", 0).remove();
  },

  clearContainer(container: Container) {
    let c = container.firstChild;
    while (c) {
      container.removeChild(c);
      c = container.firstChild;
    }
  },

  finalizeInitialChildren(
    instance: Instance,
    type: Type,
    props: any,
    rootContainer: Container,
    hostContext: any
  ) {
    return false;
  },
  commitMount(
    instance: Instance,
    type: Type,
    props: any,
    internalHandle: any
  ) {},

  getRootHostContext(rootContainer: Container) {
    return null;
  },
  getChildHostContext(
    parentHostContext: any,
    type: Type,
    rootContainer: Container
  ) {
    return parentHostContext;
  },
  getPublicInstance(instance: any) {
    return instance;
  },
  prepareForCommit(containerInfo: any) {
    return null;
  },
  resetAfterCommit(containerInfo: any) {},
  shouldSetTextContent(type: Type, props: any) {
    return false;
  },
  resetTextContent(instance: TextInstance) {},
});
