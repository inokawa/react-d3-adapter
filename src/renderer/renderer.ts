import React from "react";
import { reconciler } from "./reconciler";

const containers = new Map<HTMLElement, any>();

const render = (elem: React.ReactNode, root: HTMLElement | null): void => {
  if (!root) return;
  let container = containers.get(root);
  if (!container) {
    container = reconciler.createContainer(root, 0, false, null);
    containers.set(root, container);
  }

  reconciler.updateContainer(elem, container, null, () => undefined);
};

const unmountComponentAtNode = (root: HTMLElement) => {
  if (containers.has(root)) {
    reconciler.updateContainer(null, containers.get(root), null, () => {
      containers.delete(root);
    });
  }
};

export { render, unmountComponentAtNode };
