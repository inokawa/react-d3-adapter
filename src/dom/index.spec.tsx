import React from "react";
import renderer from "react-test-renderer";
import { D3Element } from "./";

describe("react-dom", () => {
  it("normal", () => {
    const root = new D3Element("svg", undefined, true);
    root.setAttribute("width", "600");
    root.setAttribute("height", "400");
    root.style.setProperty("width", "600px");
    root.style.setProperty("height", "400px");
    const child = new D3Element("g");
    const rect = new D3Element("rect");
    rect.setAttribute("x", "0");
    rect.setAttribute("y", "0");
    rect.setAttribute("width", "100");
    rect.setAttribute("height", "100");
    rect.setAttribute("fill", "none");
    rect.setAttribute("stroke", "red");
    rect.setAttribute("stroke-width", "1");
    const text = new D3Element("text");
    text.textContent = "test";
    root.appendChild(child);
    child.appendChild(rect);
    child.appendChild(text);
    const component = renderer.create(root.toReact());
    expect(component.toJSON()).toMatchSnapshot();
  });
});
