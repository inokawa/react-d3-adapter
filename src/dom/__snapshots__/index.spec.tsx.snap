// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-dom normal 1`] = `
<svg
  height="400"
  style={
    Object {
      "height": "400px",
      "width": "600px",
    }
  }
  width="600"
>
  <g
    style={Object {}}
  >
    <rect
      fill="none"
      height="100"
      stroke="red"
      strokeWidth="1"
      style={Object {}}
      width="100"
      x="0"
      y="0"
    />
    <text
      style={Object {}}
    >
      test
    </text>
  </g>
</svg>
`;
